src/cargo/ops/resolve.rs

```diff
@@ -213,54 +215,23 @@ pub fn resolve_with_previous<'a, 'cfg>(
         registry.lock_patches();
     }
 
-    let mut summaries = Vec::new();
     for member in ws.members() {
         registry.add_sources(&[member.package_id().source_id().clone()])?;
-        let method_to_resolve = match method {
-            // When everything for a workspace we want to be sure to resolve all
-            // members in the workspace, so propagate the `Method::Everything`.
-            Method::Everything => Method::Everything,
-
-            // If we're not resolving everything though then we're constructing the
-            // exact crate graph we're going to build. Here we don't necessarily
-            // want to keep around all workspace crates as they may not all be
-            // built/tested.
-            //
-            // Additionally, the `method` specified represents command line
-            // flags, which really only matters for the current package
-            // (determined by the cwd). If other packages are specified (via
-            // `-p`) then the command line flags like features don't apply to
-            // them.
-            //
-            // As a result, if this `member` is the current member of the
-            // workspace, then we use `method` specified. Otherwise we use a
-            // base method with no features specified but using default features
-            // for any other packages specified with `-p`.
-            Method::Required { dev_deps, .. } => {
-                let base = Method::Required {
-                    dev_deps,
-                    features: &[],
-                    all_features: false,
-                    uses_default_features: true,
-                };
-                let member_id = member.package_id();
-                match ws.current_opt() {
-                    Some(current) if member_id == current.package_id() => method,
-                    _ => {
-                        if specs.iter().any(|spec| spec.matches(member_id)) {
-                            base
-                        } else {
-                            continue;
-                        }
-                    }
-                }
-            }
-        };
-
-        let summary = registry.lock(member.summary().clone());
-        summaries.push((summary, method_to_resolve));
     }
 
+    // An edge case: user invoked something like
+    //
+    //     cargo build --package foo
+    //
+    // where `foo` is *not* a member of the workspace. We want `foo` to end up in resolve
+    // somehow, so let's resolve the whole workspace
+    let take_all = !ws.members().any(|m| requested.contains(m.package_id()));
+    let summaries = ws.members().filter(|m| take_all || requested.contains(m.package_id()))
+        .map(|member| {
+            let summary = registry.lock(member.summary().clone());
+            (summary, method)
```

Comment:
I'm having a tough time proving to myself that this loop is equivalent to what was there before. I think the idea is that if cli flags affect `method` they're forbidden earlier, right? 

That is, if `method` has `features: &["some_feature"]`, then we're guaranteed there's only one member here at this point, right? If so, can that be asserted here?

Reply:
> this loop is equivalent to what was there before

It is not *exactly* equivalent, `--all-features` and `--no-default-features` would now apply to all members of the workspace, and not just the current one. This also technically is a breaking change, and it also seems much more intuitive than the previous behavior. The reasoning about `features: &["some_feature"]` is correct, I think.

Reply:
Hm ok so something liek `cargo build --all --all-features` is now changing meaning? (just to make sure I understand)

Reply:
>Hm ok so something liek cargo build --all --all-features is now changing meaning? 

Yep!

And I think even `cargo build -p foo` can change meaning, because we no longer activate cwd package unconditionally.